### ISSUE TITLE
[Question]: Does puppeteer support firefox print pdf?

### ISSUE THREAD

**Issue by d1videbyzero** [puppeteer/puppeteer#7628]
### Bug description

Not sure if this is the right place for this but I am looking for clarification with regards to features of puppeteer (specifically with firefox). I want to use puppeteer with firefox to print some html to pdf. In the documentation, I can see that it says this functionality is unsupported and when I tested it, I just got an error and no pdf. However I came across [this](https://github.com/puppeteer/puppeteer/issues/5486) bug post and it seems to suggest there is a way to use firefox print to pdf with puppeteer.

So basically I am wondering; is print to pdf supported in puppeteer using firefox?

### Puppeteer version

10.4.0

### Node.js version

12.21.0

### npm version

7.5.2

### What operating system are you seeing the problem on?

Linux

**Comment by jschfflr** [comment]
@d1videbyzero Thanks for taking the time to report this issue! 
Could you share the error message that you got from puppeteer?

**Comment by d1videbyzero** [comment]
> @d1videbyzero Thanks for taking the time to report this issue! Could you share the error message that you got from puppeteer?

I am actually not getting any error message at all. When I set `product: 'firefox'` in the `launch()` function, the browser opens but does not react to the `page.setContent()` or `page.pdf()` functions. The browser hangs forever until I manually quit.

The reason I need firefox is for some specific quirks it has with loading some css. I am confused because in the documentation it says the `page.pdf()` function is unsupported by firefox but the issue I linked in my post suggests that it does work at least on some version. What confuses me even more is that `page.setContent()` isnt working because as far as I can tell it is supported with firefox. It may be worth noting that all of this is working as expected with chrome.

I appreciate any help.

**Comment by jschfflr** [comment]
Hmm, so it seems like the problems is actually with `page.setContent` already. 
Could you post a minimal reproduction for this situation and share the protocol messages that are sent between Puppeteer and Firefox by setting `DEBUG=puppeteer:*`? Thanks!

**Comment by d1videbyzero** [comment]
Here are the steps I am taking. Let me know if anything needs further explanation. I should also mention that this works fine when using chromium, however I seek the use of some specific qualities that firefox has when rendering content.
```
// render html
const compiledFunction = pug.compileFile('views/checklist.pug');
const htmlContent = compiledFunction(checklist);

// convert html to pdf and save file
const browser = await puppeteer.launch({ product: 'firefox' });
const page = await browser.newPage();

// Go to empty html file to get relative path for static content
const path = 'file://' + __dirname + '/../public/empty.html';
await page.goto(path);

// Set desired html content and convert to pdf
await page.setContent(htmlContent, { waitUntil: 'networkidle2' });
await page.pdf({ path: `pdfs/${req.username}.pdf`, printBackground: true });
await browser.close();
```
And here are the protocol messages I get when run like you asked.
```
Server started on port 3000...
  puppeteer:fetcher {
  puppeteer:fetcher   revision: '95.0a1',
  puppeteer:fetcher   executablePath: '/home/user/development/projects/checklist_builder_backend/checklist-service/node_modules/puppeteer/.local-firefox/linux-95.0a1/firefox/firefox',
  puppeteer:fetcher   folderPath: '/home/user/development/projects/checklist_builder_backend/checklist-service/node_modules/puppeteer/.local-firefox/linux-95.0a1',
  puppeteer:fetcher   local: true,
  puppeteer:fetcher   url: 'https://archive.mozilla.org/pub/firefox/nightly/latest-mozilla-central/firefox-95.0a1.en-US.linux-x86_64.tar.bz2',
  puppeteer:fetcher   product: 'firefox'
  puppeteer:fetcher } +0ms
  puppeteer:launcher Calling /home/user/development/projects/checklist_builder_backend/checklist-service/node_modules/puppeteer/.local-firefox/linux-95.0a1/firefox/firefox --no-remote --foreground --headless about:blank --remote-debugging-port=0 --profile /tmp/puppeteer_dev_firefox_profile-RrvxMe +0ms
  puppeteer:protocol:SEND ► {"method":"Target.setDiscoverTargets","params":{"discover":true},"id":1} +0ms
  puppeteer:protocol:RECV ◀ {"method":"Target.targetCreated","params":{"targetInfo":{"targetId":"b66c0feb-6a3d-483b-929b-07a1efec13f1","type":"page","title":"","url":"about:blank","attached":false,"browserContextId":null}}} +0ms
  puppeteer:protocol:RECV ◀ {"method":"Target.targetCreated","params":{"targetInfo":{"targetId":"743a6fe3-e138-4ee2-8f08-619ffe00569c","type":"browser","attached":true}}} +1ms
  puppeteer:protocol:RECV ◀ {"id":1,"result":{}} +0ms
  puppeteer:protocol:SEND ► {"method":"Target.createTarget","params":{"url":"about:blank"},"id":2} +22ms
  puppeteer:protocol:RECV ◀ {"method":"Target.targetCreated","params":{"targetInfo":{"targetId":"3ddb03d8-53e1-4ce0-aa18-4a928b758c41","type":"page","title":"","url":"about:blank","attached":false,"browserContextId":null}}} +14ms
  puppeteer:protocol:RECV ◀ {"id":2,"result":{"targetId":"3ddb03d8-53e1-4ce0-aa18-4a928b758c41"}} +11ms
  puppeteer:protocol:SEND ► {"method":"Target.attachToTarget","params":{"targetId":"3ddb03d8-53e1-4ce0-aa18-4a928b758c41","flatten":true},"id":3} +25ms
  puppeteer:protocol:RECV ◀ {"method":"Target.attachedToTarget","params":{"targetInfo":{"targetId":"3ddb03d8-53e1-4ce0-aa18-4a928b758c41","type":"page","title":"","url":"about:blank","attached":false,"browserContextId":null},"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","waitingForDebugger":false}} +10ms
  puppeteer:protocol:RECV ◀ {"id":3,"result":{"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0"}} +1ms
  puppeteer:protocol:SEND ► {"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","method":"Page.enable","params":{},"id":4} +12ms
  puppeteer:protocol:SEND ► {"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","method":"Page.getFrameTree","params":{},"id":5} +1ms
  puppeteer:protocol:SEND ► {"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","method":"Target.setAutoAttach","params":{"autoAttach":true,"waitForDebuggerOnStart":false,"flatten":true},"id":6} +0ms
  puppeteer:protocol:SEND ► {"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","method":"Performance.enable","params":{},"id":7} +1ms
  puppeteer:protocol:SEND ► {"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","method":"Log.enable","params":{},"id":8} +0ms
  puppeteer:protocol:RECV ◀ {"id":6,"result":{},"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0"} +22ms
  puppeteer:protocol:RECV ◀ {"method":"Target.receivedMessageFromTarget","params":{"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","message":"{\"id\":6,\"result\":{}}"}} +44ms
  puppeteer:protocol:RECV ◀ {"id":4,"result":{},"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0"} +1ms
  puppeteer:protocol:RECV ◀ {"method":"Target.receivedMessageFromTarget","params":{"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","message":"{\"id\":4,\"result\":{}}"}} +0ms
  puppeteer:protocol:RECV ◀ {"id":5,"result":{"frameTree":{"frame":{"id":"44","url":"about:blank","securityOrigin":null,"mimeType":null}}},"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0"} +0ms
  puppeteer:protocol:RECV ◀ {"method":"Target.receivedMessageFromTarget","params":{"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","message":"{\"id\":5,\"result\":{\"frameTree\":{\"frame\":{\"id\":\"44\",\"url\":\"about:blank\",\"securityOrigin\":null,\"mimeType\":null}}}}"}} +0ms
  puppeteer:protocol:RECV ◀ {"id":7,"result":{},"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0"} +0ms
  puppeteer:protocol:RECV ◀ {"method":"Target.receivedMessageFromTarget","params":{"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","message":"{\"id\":7,\"result\":{}}"}} +0ms
  puppeteer:protocol:RECV ◀ {"id":8,"result":{},"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0"} +0ms
  puppeteer:protocol:RECV ◀ {"method":"Target.receivedMessageFromTarget","params":{"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","message":"{\"id\":8,\"result\":{}}"}} +0ms
  puppeteer:protocol:SEND ► {"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","method":"Page.setLifecycleEventsEnabled","params":{"enabled":true},"id":9} +65ms
  puppeteer:protocol:SEND ► {"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","method":"Runtime.enable","params":{},"id":10} +0ms
  puppeteer:protocol:SEND ► {"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","method":"Network.enable","params":{},"id":11} +0ms
  puppeteer:protocol:RECV ◀ {"id":11,"result":{},"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0"} +3ms
  puppeteer:protocol:RECV ◀ {"method":"Target.receivedMessageFromTarget","params":{"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","message":"{\"id\":11,\"result\":{}}"}} +0ms
  puppeteer:protocol:RECV ◀ {"id":9,"result":{},"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0"} +0ms
  puppeteer:protocol:RECV ◀ {"method":"Target.receivedMessageFromTarget","params":{"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","message":"{\"id\":9,\"result\":{}}"}} +1ms
  puppeteer:protocol:RECV ◀ {"id":10,"result":{},"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0"} +0ms
  puppeteer:protocol:SEND ► {"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","method":"Page.addScriptToEvaluateOnNewDocument","params":{"source":"//# sourceURL=__puppeteer_evaluation_script__","worldName":"__puppeteer_utility_world__"},"id":12} +4ms
  puppeteer:protocol:RECV ◀ {"method":"Target.receivedMessageFromTarget","params":{"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","message":"{\"id\":10,\"result\":{}}"}} +1ms
  puppeteer:protocol:RECV ◀ {"method":"Runtime.executionContextCreated","params":{"context":{"id":1,"origin":"null","name":"","auxData":{"isDefault":true,"frameId":"44","type":"default"}}},"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0"} +0ms
  puppeteer:protocol:RECV ◀ {"method":"Target.receivedMessageFromTarget","params":{"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","message":"{\"method\":\"Runtime.executionContextCreated\",\"params\":{\"context\":{\"id\":1,\"origin\":\"null\",\"name\":\"\",\"auxData\":{\"isDefault\":true,\"frameId\":\"44\",\"type\":\"default\"}}}}"}} +1ms
  puppeteer:protocol:RECV ◀ {"id":12,"result":{"identifier":"75374bcb-eea6-4e21-84dd-85c4b4b94172"},"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0"} +1ms
  puppeteer:protocol:RECV ◀ {"method":"Target.receivedMessageFromTarget","params":{"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","message":"{\"id\":12,\"result\":{\"identifier\":\"75374bcb-eea6-4e21-84dd-85c4b4b94172\"}}"}} +0ms
  puppeteer:protocol:SEND ► {"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","method":"Page.createIsolatedWorld","params":{"frameId":"44","grantUniveralAccess":true,"worldName":"__puppeteer_utility_world__"},"id":13} +2ms
  puppeteer:protocol:RECV ◀ {"method":"Runtime.executionContextCreated","params":{"context":{"id":2,"origin":"null","name":"__puppeteer_utility_world__","auxData":{"isDefault":false,"frameId":"44","type":"isolated"}}},"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0"} +1ms
  puppeteer:protocol:RECV ◀ {"method":"Target.receivedMessageFromTarget","params":{"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","message":"{\"method\":\"Runtime.executionContextCreated\",\"params\":{\"context\":{\"id\":2,\"origin\":\"null\",\"name\":\"__puppeteer_utility_world__\",\"auxData\":{\"isDefault\":false,\"frameId\":\"44\",\"type\":\"isolated\"}}}}"}} +1ms
  puppeteer:protocol:RECV ◀ {"id":13,"result":{"executionContextId":2},"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0"} +0ms
  puppeteer:protocol:RECV ◀ {"method":"Target.receivedMessageFromTarget","params":{"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","message":"{\"id\":13,\"result\":{\"executionContextId\":2}}"}} +0ms
  puppeteer:protocol:SEND ► {"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","method":"Emulation.setDeviceMetricsOverride","params":{"mobile":false,"width":800,"height":600,"deviceScaleFactor":1,"screenOrientation":{"angle":0,"type":"portraitPrimary"}},"id":14} +2ms
  puppeteer:protocol:SEND ► {"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","method":"Emulation.setTouchEmulationEnabled","params":{"enabled":false},"id":15} +1ms
  puppeteer:protocol:RECV ◀ {"id":15,"result":{},"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0"} +3ms
  puppeteer:protocol:RECV ◀ {"method":"Target.receivedMessageFromTarget","params":{"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","message":"{\"id\":15,\"result\":{}}"}} +1ms
  puppeteer:protocol:RECV ◀ {"id":14,"result":{},"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0"} +22ms
  puppeteer:protocol:RECV ◀ {"method":"Target.receivedMessageFromTarget","params":{"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","message":"{\"id\":14,\"result\":{}}"}} +1ms
  puppeteer:protocol:SEND ► {"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","method":"Page.navigate","params":{"url":"file:///home/user/development/projects/checklist_builder_backend/checklist-service/routes/../public/empty.html","frameId":"44"},"id":16} +27ms
  puppeteer:protocol:RECV ◀ {"id":16,"result":{"frameId":"44"},"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0"} +17ms
  puppeteer:protocol:RECV ◀ {"method":"Target.receivedMessageFromTarget","params":{"sessionId":"219ee6b8-f5b4-4de3-943c-f6346d90a8d0","message":"{\"id\":16,\"result\":{\"frameId\":\"44\"}}"}} +0ms
TimeoutError: Navigation timeout of 30000 ms exceeded
    at /home/user/development/projects/checklist_builder_backend/checklist-service/node_modules/puppeteer/lib/cjs/common/LifecycleWatcher.js:101:111
  -- ASYNC --
    at Frame.<anonymous> (/home/user/development/projects/checklist_builder_backend/checklist-service/node_modules/puppeteer/lib/cjs/common/helper.js:109:19)
    at Page.goto (/home/user/development/projects/checklist_builder_backend/checklist-service/node_modules/puppeteer/lib/cjs/common/Page.js:749:53)
    at Page.<anonymous> (/home/user/development/projects/checklist_builder_backend/checklist-service/node_modules/puppeteer/lib/cjs/common/helper.js:110:27)
    at Response.<anonymous> (/home/user/development/projects/checklist_builder_backend/checklist-service/routes/checklists.js:307:36)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

**Comment by d1videbyzero** [comment]
@jschfflr Have you had a chance to look at the protocol messages?

**Comment by nurdism** [comment]
Doing a bit of debugging myself, seems to be `{ waitUntil: 'networkidle2' }` option that it is waiting on.

**Comment by OrKoN** [comment]
@whimboo do you know if printing is expected to work?

**Comment by whimboo** [comment]
As @nurdism has commented in June the problem is the usage of `waitUntil: networkidle2`. This is not supported in Firefox and causes a hang. See issue #5841, which I'm going to dupe this issue against.

Also printing to PDF works in Firefox for quite some time.
